### ISSUE TITLE
(AspNetCore) Metrics Not Registering To Same IMetricsRoot

### ISSUE THREAD

**Issue by Jon-Murray** [AppMetrics/AppMetrics#206]
Hi Al et al,

First off I just want to say I stumbled across this project a few days ago and it's absolutely fantastic, thanks. Secondly I'm having a bit of an issue with multiple sources feeding into the same IMetricsRoot.

Basically, I'm aiming to recreate the functionality for the various aspNetCore MVC middlewares, which are below:
```
            app.UseMetricsActiveRequestMiddleware();
            app.UseMetricsErrorTrackingMiddleware();
            app.UseMetricsPostAndPutSizeTrackingMiddleware();
            app.UseMetricsRequestTrackingMiddleware();
            app.UseMetricsOAuth2TrackingMiddleware();
            app.UseMetricsApdexTrackingMiddleware();
```

Now, everything is working fine for both my implementation, and the standard .net middleware. However, they will not join up. In my startup file I have the following:

```
        public Startup(IHostingEnvironment env)
        {
            StatsAndMetrics.Metrics.init();

```

Followed by:
(within ConfigureServices)
```
          services.AddMetrics(StatsAndMetrics.Metrics.metrics);
            services.AddMetricsEndpoints();
```

All StatsAndMetrics.init() does is create a static instance of an IMetricsRoot with a default builder:

  ```
      public static IMetricsRoot metrics;
        public static void init()
        {
            metrics = new MetricsBuilder()
                .Report.ToConsole()
                .Build();
```

Now, as i've mentioned before... everything works perfectly. If i make a request to my controller, and then visit the /metrics endpoint, i see the correct stats and there are no problems here. However, i'm trying to recreate this middleware for SignalR and hub invocations. I've done this by modifying the aspnetcore SignalR code so that whenever a hub is invoked, i'm basically recreating word for word what is in the middleware. For example (inside SignalR):

```
string mName = "GET " + descriptor.MethodExecutor.TargetTypeInfo.Name + "/" + invocationMessage.Target;

                List<string> tagKeys = new List<string>()
                {
                    "route",
                    "server",
                    "app",
                    "env"
                };

                List<string> tagValues = new List<string>()
                {
                    mName,
                    "JonPC",
                    "CorePort",
                    "development"
                };

                var tags = new MetricTags(tagKeys.ToArray(), tagValues.ToArray());

         var timer = StatsAndMetrics.Metrics.metrics.Provider.Timer.Instance(StatsAndMetrics.Timers.EndpointRequestTransactionDuration, tags);


                // Send the request
                await Invoke(descriptor, connection, invocationMessage);

                var elapsed = StatsAndMetrics.Metrics.metrics.Clock.Nanoseconds - startTime;


                // Record the value for this individual call.
                timer.Record(elapsed, TimeUnit.Nanoseconds);
```

Again, the above works absolutely fine. The issue is that I can't seem to get the two to marry up. When i visit /metrics, i see everything gathered from the middleware. When i manually flush my IMetricsRoot to the console (for example) I see everything from SignalR, and nothing from MVC. The measurements, tags, etc. are all the same. Below is an output from my console logging (SignalR):

```
--------------------------------------------------------------
# TIMESTAMP: 636480106884083444
# MEASUREMENT: [Application.HttpRequests] Transactions Per Endpoint
# TAGS:
                  route = GET FoldersHub/getSubNodeData
                  mtype = timer
                   unit = req
               unit_dur = ms
              unit_rate = min
# FIELDS:
            count.meter = 2
                 rate1m = 22.0810659511038
                 rate5m = 23.6033148917188
                rate15m = 23.8670363521175
              rate.mean = 11.6753610484248
                samples = 2
                   last = 29.767936
             count.hist = 2
                    sum = 107.334865
                    min = 29.767936
                    max = 77.566929
                   mean = 53.4881896370411
                 median = 29.767936
                 stddev = 23.8988243424146
                   p999 = 77.566929
                    p99 = 77.566929
                    p98 = 77.566929
                    p95 = 77.566929
                    p75 = 77.566929
--------------------------------------------------------------
```

And below is an output from /metrics (.NET/MVC):

```
# TIMESTAMP: 636480114819502345
# MEASUREMENT: [Application.HttpRequests] Transactions Per Endpoint
# TAGS:
                  route = GET project/{id?}
                 server = JonPC
                    app = CorePort
                    env = development
                  mtype = timer
                   unit = req
               unit_dur = ms
              unit_rate = min
# FIELDS:
            count.meter = 2
                 rate1m = 3.29030180732183E-05
                 rate5m = 1.612932305754
                rate15m = 9.75767183377435
              rate.mean = 0.147065702345789
                samples = 2
                   last = 15653.102366
             count.hist = 2
                    sum = 31294.959575
                    min = 15641.857209
                    max = 15653.102366
                   mean = 15647.4797875
                 median = 15653.102366
                 stddev = 5.6225785
                   p999 = 15653.102366
                    p99 = 15653.102366
                    p98 = 15653.102366
                    p95 = 15653.102366
                    p75 = 15653.102366
--------------------------------------------------------------

```

Any ideas? Nb. I can see that the "Server", "Environment", and "app" tags are missing in the SignalR implementation. I've added these in so they are exactly the same as the MVC version, but still no joy. It must be something nooby i'm doing so any advice/links to the right direction would be hugely appreciated.

Cheers!

**Comment by alhardy** [comment]
@Jon-Murray  glad you find it useful. The only think I can think of is that you are creating a second instance of `IMetrics` which is designed to be a singleton. Any chance you could push a slimmed down version of your app demonstrating the issue so I can  debug?

**Comment by Jon-Murray** [comment]
Hi, thanks for getting back to me so quickly. I'll get something together tonight or tomorrow (UK timezone) and upload it asap. 

Thanks again

**Comment by Jon-Murray** [comment]
Ok, i've made a test project which is pretty much doing what my app does. There's a readme file within the root which explains a bit about what i've done. I've cut it down as much as possible, and am still getting the same behaviour (though my instance seems only be created once).

The link is https://drive.google.com/file/d/1JFD2EJKGfB-9pVPwvKilx8s2iscuTlBY/view?usp=sharing

Thanks again for taking a look

**Comment by alhardy** [comment]
@Jon-Murray I have time again for App Metrics, before I take a look, did you end up getting to the bottom of this?

**Comment by Jon-Murray** [comment]
I couldn't figure it out, sorry
________________________________

**Comment by alhardy** [comment]
The sample is fairly messy ;) I see the problem. As I mentioned you are creating 2 instances of `IMetrics` one via the `WebHost` and the other via `Startup`. 

You don't actually need the all that metrics startup configuration you've added, simply configure you `WebHost` something like:

```

            var host = WebHost.CreateDefaultBuilder(args)
                .ConfigureMetricsWithDefaults(
                    builder =>
                    {
                        builder.Report.ToConsole(TimeSpan.FromSeconds(2));
                    })
                .UseMetrics()
                .UseStartup<Startup>()
                .Build();
```

The `HubEndPoint` is using a reference to an `IMetric` instance that you created via a static class, it's better you inject that or use the service scope factory to resolve, I've tried this and get the same instance of `IMetric` that we instantiated by the same `WebHost` configuration provided.
